gateway/src/lib.rs

```diff
@@ -683,43 +676,150 @@ impl Actor {
+                    return cross_msg.send(rt, rto);
                 }
             }
             _ => {
                 return Err(actor_error!(
                     illegal_argument,
                     "cross-message to apply dosen't have the right type"
-                ));
+                ))
             }
         };
 
-        Ok(())
+        rt.transaction(|st: &mut State, rt| {
+            let owner = cross_msg
+                .msg
+                .from
+                .raw_addr()
+                .map_err(|_| actor_error!(illegal_argument, "invalid address"))?;
+            st.insert_postbox(
+                rt.store(),
+                Some(vec![owner]),
+                MIN_CROSS_MSG_GAS.clone(),
+                cross_msg,
+            )
+            .map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "error save topdown messages")
+            })?;
+            Ok(())
+        })?;
+
+        Ok(RawBytes::default())
+    }
+
+    fn propagate<BS, RT>(rt: &mut RT, params: PropagateParams) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        // TODO: update this to EOA
+        rt.validate_immediate_caller_accept_any()?;
+
+        let PropagateParams { gas, postbox_cid } = params;
+        let owner = rt.message().caller();
+
+        rt.transaction(|st: &mut State, rt| {
+            let postbox_item = st.load_from_postbox(rt.store(), postbox_cid).map_err(|e| {
+                log::error!("encountered error loading from postbox: {:?}", e);
+                actor_error!(unhandled_message, "cannot load from postbox")
+            })?;
+
+            if let Some(owners) = postbox_item.owners && !owners.contains(&owner) {
+                return Err(actor_error!(illegal_state, "owner not match"));
+            }
+
+            if postbox_item.gas < gas {
+                return Err(actor_error!(illegal_state, "not enough gas"));
+            }
+
+            let PostBoxItem { cross_msg, .. } = postbox_item;
+            Self::commit_cross_message(rt, st, cross_msg)?;
+            st.remove_from_postbox(rt.store(), postbox_cid)
+        })?;
+
+        Ok(RawBytes::default())
+    }
+
+    /// Commit the cross message to storage.
+    ///
+    /// ALWAYS CALL THIS IN A TRANSACTION.
+    fn commit_cross_message<BS, RT>(
+        rt: &mut RT,
+        st: &mut State,
+        mut cross_msg: CrossMsg,
+    ) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let sto = cross_msg
+            .msg
+            .to
+            .subnet()
+            .map_err(|_| actor_error!(illegal_argument, "error getting subnet from msg"))?;
+        if sto == st.network_name {
+            return Err(actor_error!(illegal_state, "should already be committed"));
+        }
+
+        // TODO: runtime should expose is in transaction to ensure in transaction check
```

Comment:
I don't understand this comment. Do you need any clarification or should we remove it?

Reply:
I was thinking of to programatically enforce this, instead of in comments.

Reply:
We can remove this comment for now